Content.Server/Construction/MachineFrameSystem.cs

```diff
@@ -0,0 +1,298 @@
+        {
+            component.ComponentProgress[compName] = 0;
+        }
+
+        foreach (var (compName, _) in component.TagRequirements)
+        {
+            component.TagProgress[compName] = 0;
+        }
+    }
+
+    public void RegenerateProgress(MachineFrameComponent component)
+    {
+        AppearanceComponent? appearance;
+
+        if (!component.HasBoard)
+        {
+            if (TryComp(component.Owner, out appearance)) appearance.SetData(MachineFrameVisuals.State, 1);
+
+            component.TagRequirements.Clear();
+            component.MaterialRequirements.Clear();
+            component.ComponentRequirements.Clear();
+            component.TagRequirements.Clear();
+            component.Progress.Clear();
+            component.MaterialProgress.Clear();
+            component.ComponentProgress.Clear();
+            component.TagProgress.Clear();
+
+            return;
+        }
+
+        var board = component.BoardContainer.ContainedEntities[0];
+
+        if (!TryComp<MachineBoardComponent>(board, out var machineBoard))
+            return;
+
+        if (TryComp(component.Owner, out appearance)) appearance.SetData(MachineFrameVisuals.State, 2);
+
+        ResetProgressAndRequirements(component, machineBoard);
+
+        foreach (var part in component.PartContainer.ContainedEntities)
+        {
+            if (TryComp<MachinePartComponent>(part, out var machinePart))
+            {
+                // Check this is part of the requirements...
+                if (!component.Requirements.ContainsKey(machinePart.PartType))
+                    continue;
+
+                if (!component.Progress.ContainsKey(machinePart.PartType))
+                    component.Progress[machinePart.PartType] = 1;
+                else
+                    component.Progress[machinePart.PartType]++;
+            }
+
+            if (TryComp<StackComponent>(part, out var stack))
+            {
+                var type = stack.StackTypeId;
+                // Check this is part of the requirements...
+                if (!component.MaterialRequirements.ContainsKey(type))
+                    continue;
+
+                if (!component.MaterialProgress.ContainsKey(type))
+                    component.MaterialProgress[type] = 1;
+                else
+                    component.MaterialProgress[type]++;
+            }
+
+            // I have many regrets.
+            foreach (var (compName, _) in component.ComponentRequirements)
+            {
+                var registration = _factory.GetRegistration(compName);
+
+                if (!HasComp(part, registration.Type))
+                    continue;
+
+                if (!component.ComponentProgress.ContainsKey(compName))
+                    component.ComponentProgress[compName] = 1;
+                else
+                    component.ComponentProgress[compName]++;
+            }
+
+            // I have MANY regrets.
```

Comment:
thank you for keeping these comments, they really add to the experience :relieved: